postgres/datadog_checks/postgres/connections.py

```diff
@@ -0,0 +1,102 @@
+import datetime
+import inspect
+import threading
+from collections import namedtuple
+
+import psycopg2
+
+ConnectionWithTTL = namedtuple("ConnectionWithTTL", "connection deadline")
+
+
+class MultiDatabaseConnectionPool(object):
+    """
+    Manages a connection pool of connections across many logical databases with a maximum
+    of 1 conn per database. Traditional connection pools manage a set of connections to a
+    single database, however the usage patterns of the Agent application should aim to have
+    minimal footprint and reuse a single connection as much as possible.
+
+    Even when limited to a single connection per database, an instance with hundreds of
+    databases still present a connection overhead risk. This class provides a mechanism
+    to prune connections to a database which were not used in the time specified by their
+    TTL.
+    """
+
+    class Stats(object):
+        def __init__(self):
+            self.connection_opened = 0
+            self.connection_pruned = 0
+            self.connection_closed = 0
+            self.connection_closed_failed = 0
+
+        def __repr__(self):
+            return str(self.__dict__)
+
+        def reset(self):
+            self.__init__()
+
+    def __init__(self, connect_fn):
+        self._stats = self.Stats()
+        self._mu = threading.Lock()
```

Comment:
When reading how this class is used, I don't see any multi-threading scenarios. Am I missing something?

Reply:
For now the pool is only used in one threaded class for DBMAsyncJob. I would ultimately like to use it across the DBMAsyncJobs and the main postgres integration. But that's a longer-term change. The multithreaded safety is put in place now so as not to lock in any changes that will have to be undone later.

Reply:
oki-doki, let's hope this isn't premature optimization ;)